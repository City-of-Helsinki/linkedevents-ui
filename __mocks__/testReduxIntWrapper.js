import {IntlProvider, addLocaleData} from 'react-intl'
import {MemoryRouter as Router} from 'react-router'
import {Provider} from 'react-redux'
import React from 'react'
//Changed Intl-providers testmessage
import fiMessages from 'src/i18n/fi.json';
import mapValues from 'lodash/mapValues';
const testMessages = mapValues(fiMessages, (value, key) => value);
const ReduxIntWrappeComponent = (store, component, locale = 'fi') => {
    const lang = require('react-intl/locale-data/' + locale)
    addLocaleData(lang)
    return (
        <Provider store={store}>
            <IntlProvider locale={locale} messages={testMessages}>
                <Router>
                    {component}
                </Router>
            </IntlProvider>
        </Provider>
    )
}
export default ReduxIntWrappeComponent;
