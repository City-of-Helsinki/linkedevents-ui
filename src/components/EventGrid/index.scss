.event-item {
    position: relative;

    height: 228px;

    margin-top: 15px;
    margin-bottom: 15px;
    overflow: hidden;

    .thumbnail {
        // position: absolute;
        // top: 0;
        // left: 0;

        width: 100%;
        height: 100%;

        background-color: #0072c6;

        background-size: cover;
        background-repeat: no-repeat;
        background-position: 50% 50%;
    }

    .name {
        font-family: 'Roboto', sans-serif;

        display: flex;
        align-items: center;
        justify-content: flex-start;

        color: #ffffff;
        font-size: 17px;
        font-weight: 400;

        position: absolute;
        bottom: 0;
        left: 0;

        height: 90px;
        width: 100%;

        padding-left: 22px;

        background: rgba(0,0,0,0.5);
    }

    .converted-day {
        padding-right: 10px;
    }

    .search-badge {
        float: left;
        margin: 2px;
        text-transform: uppercase;
    }
}
