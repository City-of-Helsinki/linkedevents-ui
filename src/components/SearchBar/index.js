import './index.scss';

import React, {useState} from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import {FormattedMessage, injectIntl} from 'react-intl';
import CustomDatePicker from '../CustomFormFields/CustomDatePicker'
import {Button, Form, FormGroup} from 'reactstrap';

const handleKeyPress = (event, startDate, endDate, onFormSubmit, setSearchQuery) => {
    if (event.key === 'Enter') {
        const searchQuery = event.target.value;

        onFormSubmit(searchQuery, startDate, endDate);
        setSearchQuery(searchQuery);
    }
};

const SearchBar = ({intl, onFormSubmit}) => {
    const [startDate, setStartDate] = useState(moment().startOf('day'));
    const [endDate, setEndDate] = useState(null);
    const [searchQuery, setSearchQuery] = useState('');

    return (
        <div className='search-bar'>
            <div className='search-bar--dates'>
                <p className='search-bar--label'>
                    <FormattedMessage id='pick-time-range' />
                </p>
                <CustomDatePicker
                    id="startTime"
                    name="startTime"
                    label="search-date-label-start"
                    defaultValue={startDate}
                    onChange={setStartDate}
                    maxDate={endDate ? endDate : undefined}
                    type="date"
                />
                <CustomDatePicker 
                    id="endTime"
                    name="endTime"
                    label="search-date-label-end"
                    defaultValue={endDate}
                    onChange={setEndDate}
                    minDate={startDate ? startDate : undefined}
                    type="date"
                />
            </div>
            <div className='search-bar--input event-input'>
                <Form>
                    <input
                        aria-hidden='true'
                        className='hidden'
                        onInput={(e) => setSearchQuery(e.target.value)}
                        onKeyPress={(e) =>
                            handleKeyPress(e, startDate, endDate, onFormSubmit, setSearchQuery)
                        }
                    />
                    <FormGroup>
                        <label>{intl.formatMessage({id: 'event-name-or-place'})}</label>
                        <input
                            className='event-search-bar'
                            type='text'
                            autoFocus
                            onChange={(e) => setSearchQuery(e.target.value)}
                            onKeyPress={(e) =>
                                handleKeyPress(e, startDate, endDate, onFormSubmit, setSearchQuery)
                            }
                        />
                    </FormGroup>
                </Form>
                <Button
                    disabled={searchQuery.length === 0}
                    variant='contained'
                    color='primary'
                    onClick={() => onFormSubmit(searchQuery, startDate, endDate)}>
                    <FormattedMessage id='search-event-button' />
                </Button>
            </div>
        </div>
    );
};

SearchBar.propTypes = {
    onFormSubmit: PropTypes.func,
    intl: PropTypes.object,
};

export default injectIntl(SearchBar);
