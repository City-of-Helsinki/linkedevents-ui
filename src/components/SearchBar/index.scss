.search-bar {
    &--dates {
        align-items: baseline;
        display: inline-flex;
    }

    &--label {
        margin-right: 2rem;
    }

    &--icon {
        margin: 0 16px;
        transform: translateY(8px);
    }

    &--input {
        align-items: center;
        display: flex;
        margin: 16px 0 32px;
        width: 100%;

        > div {
            flex: 1 1 auto;
        }

        > button {
            flex: 1 0 auto;
        }
    }
}

@media (max-width: 600px) {
    .search-bar--dates {
        flex-direction: column;
    }
}


@media (min-width: 992px) {
    .search-bar--input {
        width: 70%;
    }
}
