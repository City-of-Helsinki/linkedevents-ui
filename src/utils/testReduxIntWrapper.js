import { IntlProvider } from 'react-intl'
import { MemoryRouter as Router } from 'react-router'
import { Provider } from 'react-redux'
import React from 'react'
import translations from '../i18n'

export default (store, component, locale = 'fi') => {
  return (
    <Provider store={store}>
      <IntlProvider locale={locale} messages={translations[locale]}>
        <Router>
            {component}
        </Router>
      </IntlProvider>
    </Provider>
  )
}
