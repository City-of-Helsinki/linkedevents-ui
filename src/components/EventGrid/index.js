import React from 'react'
import {Link} from 'react-router-dom'
import {connect} from 'react-redux'
import moment from 'moment'
import PropTypes from 'prop-types'
import {FormattedMessage} from 'react-intl'
import get from 'lodash/get'
import forEach from 'lodash/forEach'
import {getBadge} from '../../utils/helpers'

import defaultThumbnail from '../../assets/images/helsinki-coat-of-arms-white.png'

import constants from '../../constants'
import './index.scss'

const dateFormat = function (timeStr) {
    return timeStr ? moment(timeStr).format('YYYY-MM-DD') : ''
}

const getName = (event) => {
    const paths = [
        'name.fi',
        'name.en',
        'name.sv',
        'headline.fi',
        'headline.en',
        'headline.sv',
    ]

    let name
    forEach(paths, (path) => {
        name = get(event, path)

        // We found a name, exit forEach
        if (name) {
            return false
        }
    })

    if (name) {
        return name
    }

    // If we don't have a name at this point, try to find a translation/locale from name field.
    forEach(get(event, 'name', []), (item) => {
        name = item
        return false
    })

    if (name) {
        return name
    }

    // If name field didn't have any translation/locale, then try to find one in headline field.
    forEach(get(event, 'headline', []), (item) => {
        name = item
        return false
    })

    // If name still doesn't have a value, then return an empty string.
    return name ? name : ''
}

const EventItem = (props) => {
    const name = getName(props.event)
    const url = '/event/' + encodeURIComponent(props.event['id']) + '/'
    const image = props.event.images.length > 0 ? props.event.images[0].url : defaultThumbnail
    const thumbnailStyle = {
        backgroundImage: 'url(' + image + ')',
    }
    const getDay = props.event.start_time
    let convertedDate = ''
    if (getDay) {
        const date = getDay.split('T')
        convertedDate = date[0].split('-').reverse().join('.')
    }
    const isCancelled = props.event.event_status === constants.EVENT_STATUS.CANCELLED
    const isPostponed = props.event.event_status === constants.EVENT_STATUS.POSTPONED

    return (
        <div className="col-xs-12 col-md-6 col-lg-4" key={props.event['id']}>
            <Link to={url}>
                <div className="event-item">
                    {isCancelled && getBadge('cancelled')}
                    {isPostponed && getBadge('postponed')}
                    <div className="thumbnail" style={thumbnailStyle}/>
                    <div className="name">
                        <span className="converted-day">{convertedDate}</span>
                        {name}
                    </div>

                </div>
            </Link>
        </div>
    )
}

const EventGrid = (props) => (
    <div className="row event-grid">
        {props.events.map(event => <EventItem event={event} key={event.id}/>)}
    </div>
)

EventItem.propTypes = {
    event: PropTypes.object,
}

EventGrid.propTypes = {
    events: PropTypes.array,
}

export default connect()(EventGrid)
