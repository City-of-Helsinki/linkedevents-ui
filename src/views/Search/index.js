import './index.scss'

import React from 'react'
import {FormattedMessage} from 'react-intl'
import PropTypes from 'prop-types'
import EventGrid from '../../components/EventGrid'
import SearchBar from '../../components/SearchBar'
import {EventQueryParams, fetchEvents} from '../../utils/events'
//Replaced Material-ui Spinner for a Bootstrap implementation. - Turku
import Spinner from 'react-bootstrap/Spinner'

class SearchPage extends React.Component {

    state = {
        events: [],
        loading: false,
        searchExecuted: false,
    }

    searchEvents = async (searchQuery, startDate, endDate) => {
        if (!searchQuery && (!startDate || !endDate)) {
            return
        }

        this.setState({loading: true})

        const queryParams = new EventQueryParams()
        queryParams.page_size = 100
        queryParams.sort = 'start_time'
        queryParams.nocache = Date.now()
        queryParams.text = searchQuery
        if (startDate) queryParams.start = startDate.format('YYYY-MM-DD')
        if (endDate) queryParams.end = endDate.format('YYYY-MM-DD')

        try {
            const response = await fetchEvents(queryParams)
            this.setState({events: response.data.data, searchExecuted: true})
        } finally {
            this.setState({loading: false})
        }
    }

    getResults = () => {
        const {searchExecuted, events} = this.state

        return searchExecuted && !events.length > 0
            ? <div className="search-no-results"><FormattedMessage id="search-no-results"/></div>
            : <EventGrid events={events} />
    }

    render() {
        const {loading} = this.state

        return (
            <div className="container">
                <h1><FormattedMessage id={`search-${appSettings.ui_mode}`}/></h1>
                <p><FormattedMessage id="search-events-description"/></p>
                <SearchBar onFormSubmit={(query, start, end) => this.searchEvents(query, start, end)}/>
                {loading
                    ? <div className="search-loading-spinner"><Spinner animation="border" role="status">
                        <span className="sr-only">Loading...</span>
                    </Spinner> </div>
                    : this.getResults()
                }
            </div>
        )
    }
}

SearchPage.propTypes = {
    events: PropTypes.array,
    loading: PropTypes.bool,
    searchExecuted: PropTypes.bool,
}

export default SearchPage
