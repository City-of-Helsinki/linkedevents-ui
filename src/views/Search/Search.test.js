import configureStore from 'redux-mock-store'
import React from 'react'
import thunk from 'redux-thunk'

import testReduxIntWrapper from '../../../__mocks__/testReduxIntWrapper'
import Search from './index'
import SearchBar from 'src/components/SearchBar/index'
import EventGrid from 'src/components/EventGrid';
import {mockUser} from '__mocks__/mockData';
import {shallow} from 'enzyme'
import Spinner from 'react-bootstrap/Spinner'
import {FormattedMessage} from 'react-intl';

const mockStore = configureStore([thunk]);
const initialStore = {
    user: mockUser,
    events: {
        apiErrorMsg: null,
        isFetching: false,
        fetchComplete: false,
        items: [],
        eventError: null,
        eventsError: null,
    },
};

// mock moment to render search dates as expected
jest.mock('moment');

describe('Search Snapshot', () => {
    let store;

    it('should render view correctly', () => {
        const componentProps = {
            match: {
                params: {
                    action: 'search',
                },
            },
        } // Props which are added to component
        store = mockStore(initialStore);
        const componentToTest = <Search {...componentProps} />
        const wrapper = shallow(testReduxIntWrapper(store, componentToTest))

        expect(wrapper).toMatchSnapshot()

    })
})
const defaultProps = {

}
describe('Search', () => {
    function getWrapper(props) {
        return shallow(<Search {...defaultProps} {...props}/>);
    }
    describe('render', () => {

        test('SearchBar element', () => {
            const element = getWrapper().find(SearchBar);
            expect(element).toHaveLength(1);
            expect(element.props('onFormSubmit')).toBeDefined();
        });

        test('Spinner element when loading', () => {
            const element = getWrapper();
            element.setState({loading: true});
            const spinnerElement = element.find(Spinner);
            expect(spinnerElement).toHaveLength(1);
        });

        describe('FormattedMessages', () => {
            test('default amount', () => {
                const element = getWrapper().find(FormattedMessage);
                expect(element).toHaveLength(3);
            });
            test('correct amount when !events & searchExecuted', () => {
                const element = getWrapper();
                element.setState({searchExecuted: true});
                const formattedMessages = element.find(FormattedMessage);
                expect(formattedMessages).toHaveLength(4);
            });
            test('correct amount when events & searchExecuted', () => {
                const element = getWrapper()
                element.setState({searchExecuted: true, events:['Tammi','Koivu']});
                const grid = element.find(EventGrid)
                expect(grid).toHaveLength(1);
                expect(grid.prop('events')).toEqual(element.state('events'));
            });
        });

        describe('correct count to search results FormattedMessage', () => {
            let element;
            
            beforeEach(() => {
                element = getWrapper();
            });
            test('when no results', () => {
                const resultCount = element.find(FormattedMessage).at(2);
                const count = element.state('events');
                expect(resultCount.prop('values')).toEqual({'count': count.length});
            });
            test('when results', () => {
                element.setState({events:['Tammi','Koivu']});
                const resultCount = element.find(FormattedMessage).at(2);
                const count = element.state('events');
                expect(resultCount.prop('values')).toEqual({'count': count.length});
            });
        });
    });
})



