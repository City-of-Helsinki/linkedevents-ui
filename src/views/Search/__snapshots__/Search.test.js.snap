// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Snapshot should render view correctly 1`] = `
<div
  className="container"
>
  <h1>
    <span>
      Etsi tapahtumia
    </span>
  </h1>
  <p>
    <span>
      Katsele kaikkia Linked Eventsin tapahtumia.
    </span>
  </p>
  <form
    className="row search-bar"
    onSubmit={[Function]}
  >
    <div
      className="col-sm-2 time-label"
    >
      <label>
        <span>
          Etsi vain ajalta
        </span>
      </label>
    </div>
    <div
      className="col-xs-8 col-sm-5 col-md-3 start-date"
    >
      <div
        className="hel-text-field"
      >
        <div
          className="react-datepicker-wrapper"
        >
          <div
            className="react-datepicker__input-container"
          >
            <input
              autoComplete={undefined}
              autoFocus={undefined}
              className=""
              disabled={false}
              id={undefined}
              name="startDate"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              placeholder="pp.kk.vvvv"
              readOnly={false}
              required={undefined}
              tabIndex={undefined}
              title={undefined}
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="col-xs-8 col-sm-5 col-md-3"
    >
      <div
        className="hel-text-field"
      >
        <div
          className="react-datepicker-wrapper"
        >
          <div
            className="react-datepicker__input-container"
          >
            <input
              autoComplete={undefined}
              autoFocus={undefined}
              className=""
              disabled={false}
              id={undefined}
              name="endDate"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              placeholder="pp.kk.vvvv"
              readOnly={false}
              required={undefined}
              tabIndex={undefined}
              title={undefined}
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="col-sm-8 col-xs-12"
    >
      <div
        className="text-field"
      >
        <label
          className="control-label"
          htmlFor={undefined}
        >
          <span>
            Tapahtuman nimi tai paikka
          </span>
        </label>
        <input
          autoFocus={true}
          className="form-control"
          id={undefined}
          onChange={[Function]}
          placeholder="Tapahtuman nimi tai paikka"
          type={undefined}
        />
      </div>
    </div>
    <div
      className="col-sm-4 col-xs-12"
    >
      <button
        className="mui-raised-button btn btn-default"
        color="primary"
        disabled={false}
        onClick={[Function]}
        style={
          Object {
            "height": "72px",
          }
        }
        type="submit"
      >
        <span>
          Hae tapahtumia
        </span>
      </button>
    </div>
    <p />
  </form>
  <div
    className="loadedContent"
  >
    <div
      className="row event-grid"
    />
  </div>
</div>
`;
